freedomofpress/securedrop-client
# Improve download file handling

# Description

Eliminate `File.original_file` column and hard links to downloaded files. Now as files are downloaded and decrypted, `File.filename` will be updated.

Instead of downloading directly into the root of the data directory, create a directory therein for each downloadable object, named with its UUID, and process downloaded files in that directory. This should
eliminate collisions if multiple sources upload documents with the same names.

Fixes #714.

# Test Plan

- Start a SecureDrop development server with `make dev`
- Visit the source interface and upload a file.
- Run the client with `run.sh`. Find your source in the list and:
  - download the file
  - verify that it can be opened in a disp VM by clicking on it
  - verify that it can be exported
  - verify that a metadata sync does not change the filename
- In the shell, confirm that exactly one file is present under the `$SDC_HOME/data_dir/File/{file-uuid}/` directory, with no files hard linked to it.

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:


If these changes add or remove files other than client code, packaging logic (e.g., the AppArmor profile) may need to be updated. Please check as applicable:
